### PR TITLE
[tlse] tls for PlacementAPI pod configuration

### PR DESCRIPTION
**Pull request by Deydra71**
Public/Internal service cert secrets and the CA bundle secret can be passed to configure httpd virtual hosts for tls termination. The certs get direct mounted to the appropriate place in etc/pki/tls/certs/%s.crt|key and a CA bundle to
/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem . Job deployments for bootstrap/cron get the CA bundle added if configured.

Depends-On: https://github.com/openstack-k8s-operators/lib-common/pull/428

Jira: [OSPRH-2368](https://issues.redhat.com/browse/OSPRH-2368)

### CODE REVIEW

File: templates/placementapi/config/placement-api-config.json
Comment (Deydra71) [inline templates/placementapi/config/placement-api-config.json]: @stuggi 
Since we will be using one CA bundle file created by openstack-operator, we can specify the individual file instead  the general folder (?)

**Comment by SeanMooney** [comment]
this PR should contian updates to the samples and kuttl and even tests to test it
before it can be merged

**Comment by Deydra71** [comment]
> this PR should contian updates to the samples and kuttl and even tests to test it before it can be merged

It's in the draft state still ^^

**Comment by SeanMooney** [comment]
yes i know just wanted to call that out
apparently the patches that added the service override functionality did not add that coverage so i don't want use  to continue to extend this functionality uhntil that gap is closed

File: templates/placementapi/config/placement-api-config.json
Comment (stuggi) [reply templates/placementapi/config/placement-api-config.json]: yes lets do that

**Comment by stuggi** [comment]
> yes i know just wanted to call that out apparently the patches that added the service override functionality did not add that coverage so i don't want use to continue to extend this functionality uhntil that gap is closed

I had added envtest coverage in serviceoverride PR https://github.com/openstack-k8s-operators/placement-operator/blob/main/tests/functional/placementapi_controller_test.go#L452 / https://github.com/openstack-k8s-operators/placement-operator/blob/main/tests/functional/placementapi_controller_test.go#L537

**Review by gibizer — COMMENTED** [review]
The general direction looks good to me

**Comment by Deydra71** [comment]
@gibizer  Could you please review, and give ok to test label if sufficient? I still need to make updates to kuttl and functional test, but otherwise the general TLS support is set from major part.

File: api/bases/placement.openstack.org_placementapis.yaml
Comment (stuggi) [inline api/bases/placement.openstack.org_placementapis.yaml]: I just realized that this is probably not enough. we create two services per api, public and internal to allow to use different service types (also LoadBalancer types if you do not use routes for public). With the current tls struct here, we only allow passing in one secret for the service operator.

Comment (stuggi) [reply api/bases/placement.openstack.org_placementapis.yaml]: I mean this is not a blocker for this PR since its main focus is to get the cacert bundle in so that it can validate e.g.  tls connection to db/rabbit. the second step would be to enable the tls for placement itself. just something we have to enhance.

Comment (Deydra71) [reply api/bases/placement.openstack.org_placementapis.yaml]: I see, updating the Service struct could bring more mess as it's already being used in other's implementations. I suggest keeping this PR focused on CA bundle support and then in next implement the support for the public and internal service type.

**Comment by Deydra71** [comment]
On hold until the agreed changes of the TLS structure are implemented in the lib-common.

File: api/go.mod
Comment (stuggi) [inline api/go.mod]: can you update to the one which is from your lib-common PR and rerun `make manifests`

File: go.mod
Comment (stuggi) [inline go.mod]: same

File: pkg/placement/dbsync.go
Comment (stuggi) [inline pkg/placement/dbsync.go]: should be
`volumes = append(volumes, instance.Spec.TLS.CreateVolume())` instead of calling the func again.

same for the bellow lines

File: pkg/placement/deployment.go
Comment (stuggi) [inline pkg/placement/deployment.go]: same as mentioned above. no need to call the funcs again. just, append the tls stuff

File: pkg/placement/deployment.go
Comment (stuggi) [inline pkg/placement/deployment.go]: I think this is not wanted. this would revert things from https://github.com/openstack-k8s-operators/placement-operator/commit/8fdff269484cd58ca9271f44b785a48647299f12

File: tests/functional/base_test.go
Comment (stuggi) [inline tests/functional/base_test.go]: maybe rename the const and use them here as well and not just in the GetTLSPlacementAPISpec?

File: tests/functional/suite_test.go
Comment (stuggi) [inline tests/functional/suite_test.go]: already there in line 140

File: config/samples/placement_v1beta1_placementapi.yaml
Comment (SeanMooney) [inline config/samples/placement_v1beta1_placementapi.yaml]: instead of commeting this out we should provide a separate tlse example file

File: pkg/placement/initcontainer.go
Comment (SeanMooney) [inline pkg/placement/initcontainer.go:1]: we evenutally want ot remove all init container form this operator so can we avoid adding new usage of them?

Comment (SeanMooney) [reply pkg/placement/initcontainer.go:1]: https://github.com/openstack-k8s-operators/placement-operator/issues/64

File: pkg/placement/volumes.go
Comment (stuggi) [inline pkg/placement/volumes.go]: needs to be reverted as there are two kolla configs, one for the dbsync and one for the api

File: pkg/placement/dbsync.go
Comment (stuggi) [inline pkg/placement/dbsync.go]: should be `getVolumeMounts("dbsync")`

File: pkg/placement/deployment.go
Comment (stuggi) [inline pkg/placement/deployment.go]: should be `getVolumeMounts("api")`

File: controllers/placementapi_controller.go
Comment (gibizer) [inline controllers/placementapi_controller.go:350]: I'm wondering if we can do some generic code that uses the constant field selectors (like passwordSecretField)  to access the field. So instead of having 4 blocks copy pasted here for the 4 secret, we could have a single loop that loops over all the field selectors. This can be considered in a follow up.

File: controllers/placementapi_controller.go
Comment (gibizer) [inline controllers/placementapi_controller.go:904]: I think we should set the InputReadyCondition here to False with a message that point out which input is missing. Otherwise it is hard to tell from the CR status why the deplyoment is not progressing.

File: controllers/placementapi_controller.go
Comment (gibizer) [inline controllers/placementapi_controller.go:923]: ditto.

File: controllers/placementapi_controller.go
Comment (gibizer) [inline controllers/placementapi_controller.go:1019]: +1 thanks for propagating the error to the status.

File: templates/placementapi/config/httpd.conf
Comment (gibizer) [inline templates/placementapi/config/httpd.conf:32]: nit: is this some debugging leftover?

**Review by gibizer — CHANGES_REQUESTED** [review]
I only looked at the implementation I still have to look at the test code. I request changes in this PR only for the InputReady condition setting. My refactoring suggestions totally OK to be considered for a follow up.

File: controllers/placementapi_controller.go
Comment (stuggi) [reply controllers/placementapi_controller.go:350]: I previously had a quick look to use something like `reflect.Indirect(v).FieldByName(field)` but it did not work so we considered that a later optimization work to look into. agreed to check this in a follow up

File: controllers/placementapi_controller.go
Comment (stuggi) [reply controllers/placementapi_controller.go:904]: right, makes sense

File: templates/placementapi/config/httpd.conf
Comment (stuggi) [reply templates/placementapi/config/httpd.conf:32]: this is intentional to have a comment in the httpd.conf for the virtual host. if it is considered to be useless we can remove it

File: controllers/placementapi_controller.go
Comment (stuggi) [reply controllers/placementapi_controller.go:904]: @Deydra71 check my keystone pr, there I had already added a TLSInputReady condition. lets also get this into the PR

File: templates/placementapi/config/httpd.conf
Comment (gibizer) [reply templates/placementapi/config/httpd.conf:32]: Ahh, at first I thought that this is a comment in the template, but actually it is a comment template that is expaneded by the config generation to get a comment in the resulting config file. So this is good. Keep it. :)

File: controllers/placementapi_controller.go
Comment (gibizer) [reply controllers/placementapi_controller.go:904]: Thanks @Deydra71 for the fix.

File: tests/functional/base_test.go
Comment (gibizer) [inline tests/functional/base_test.go]: you can use `names.InternalCertSecretName` and the other refs from the Names struct here. There is always a global names variable with the naming constants grouped together (it is created in suite_test). If you use `names` here then you can inline the test global `InternalCertSecretName` into CreateNames(). I know that this sounds like meaningless busy work and for a true constant like how InternalCertSecretName is defined now it probably is. The reason we have the Names struct is to encapsulate generated names like Role and RoleBinding. (nova-operator has a bunch more examples) to avoid replicating the generation logic across multiple test case.

Anyhow I let you decide which way to go as it is not a big deal.

File: tests/functional/placementapi_controller_test.go
Comment (gibizer) [inline tests/functional/placementapi_controller_test.go:1]: Thanks you for adding env test coverage. I much appreciate the effort. Also the test looks good!

File: tests/kuttl/common/tls_certificates.yaml
Comment (gibizer) [inline tests/kuttl/common/tls_certificates.yaml:1]: I agree that this is a good compromise to keep the test independent.

File: tests/kuttl/tests/placement_deploy_tls/03-assert.yaml
Comment (gibizer) [inline tests/kuttl/tests/placement_deploy_tls/03-assert.yaml:85]: the new TLSInputReady condition is missing from here

**Review by gibizer — APPROVED** [review]
This PR looks solid. The kuttl test needs a fix but other than that this is good to land from my perspective. Thanks for the effort!

File: tests/functional/base_test.go
Comment (Deydra71) [reply tests/functional/base_test.go]: I see, thanks for the explanation! Will edit it, to preserve same logic within tests.

File: controllers/placementapi_controller.go
Comment (SeanMooney) [reply controllers/placementapi_controller.go:350]: ok so i agree this can  be refactored to reduce the duplications

the lamda/funciton passed to IndexField  is just returning a StringSlice/charArray or nil if the filed is not set.

we know the mapphng between the string constants and the struct fields so we just need a pointer to member data for each and store a map of "string" -> "pointer to member data" which we can apply the the spec object to select the data 

if this was c++ that would be simple in golang I'm not sure that this easlisly do able so a slight different approach is needed 
```
field_indexers := map[string]interface{}{
    passwordSecretField: func(rawObj client.Object) []string {
		// Extract the secret name from the spec, if one is provided
		cr := rawObj.(*placementv1.PlacementAPI)
		if cr.Spec.Secret == "" {
			return nil
		}
		return []string{cr.Spec.Secret}
	},
	caBundleSecretNameField: func(rawObj client.Object) []string {
		// Extract the secret name from the spec, if one is provided
		cr := rawObj.(*placementv1.PlacementAPI)
		if cr.Spec.TLS.CaBundleSecretName == "" {
			return nil
		}
		return []string{cr.Spec.TLS.CaBundleSecretName}
	},
	tlsAPIInternalField: func(rawObj client.Object) []string {
		// Extract the secret name from the spec, if one is provided
		cr := rawObj.(*placementv1.PlacementAPI)
		if cr.Spec.TLS.API.Internal.SecretName == nil {
			return nil
		}
		return []string{*cr.Spec.TLS.API.Internal.SecretName}
	},
	tlsAPIPublicField: func(rawObj client.Object) []string {
		// Extract the secret name from the spec, if one is provided
		cr := rawObj.(*placementv1.PlacementAPI)
		if cr.Spec.TLS.API.Public.SecretName == nil {
			return nil
		}
		return []string{*cr.Spec.TLS.API.Public.SecretName}
	}
}

for field, indexer := range field_indexers{
    if err := mgr.GetFieldIndexer().IndexField(context.Background(), &placementv1.PlacementAPI{}, field, indexer
);err != nil {
		return err
	}

}
```

i think that is the same and will work but its not much simpler

the functor objects/lamdas are basically just c++ pointer to data members

Comment (SeanMooney) [reply controllers/placementapi_controller.go:350]: i have an example of this patteren here

https://github.com/openstack-k8s-operators/nova-operator/blob/8f204945b887a562229143f40901c4540925e9cd/controllers/common.go#L294-L341

File: tests/kuttl/common/tls_certificates.yaml
Comment (bogdando) [inline tests/kuttl/common/tls_certificates.yaml]: indent

File: tests/kuttl/common/tls_certificates.yaml
Comment (bogdando) [inline tests/kuttl/common/tls_certificates.yaml]: indent

File: tests/kuttl/common/tls_certificates.yaml
Comment (bogdando) [inline tests/kuttl/common/tls_certificates.yaml]: indent

File: tests/kuttl/common/tls_certificates.yaml
Comment (bogdando) [inline tests/kuttl/common/tls_certificates.yaml]: ditto

File: controllers/placementapi_controller.go
Comment (bogdando) [inline controllers/placementapi_controller.go:359]: could you please query this once (the whole spec) and store in a local var? Then you can fetch by needed fields

File: controllers/placementapi_controller.go
Comment (bogdando) [inline controllers/placementapi_controller.go:1012]: ``deplDef, err =`` ?

File: controllers/placementapi_controller.go
Comment (gibizer) [inline controllers/placementapi_controller.go:415]: This should be added as a Watch in via the NewControllerManagedBy(mgr) builder otherwise the TLS secrets are not watched.

```golang
		Watches(
			&source.Kind{Type: &corev1.Secret{}},
			handler.EnqueueRequestsFromMapFunc(r.findObjectsForSrc),
			builder.WithPredicates(predicate.ResourceVersionChangedPredicate{}),
		).
```

Comment (gibizer) [reply controllers/placementapi_controller.go:415]: Here is a commit with a test case that shows that the secret watching does not work without it https://github.com/gibizer/placement-operator/commit/f8ae8eea22a204157f84c779facda9e9d6d12533

**Review by gibizer — CHANGES_REQUESTED** [review]

File: controllers/placementapi_controller.go
Comment (Deydra71) [reply controllers/placementapi_controller.go:1012]: Not sure what you mean, can you please elaborate?

File: controllers/placementapi_controller.go
Comment (Deydra71) [reply controllers/placementapi_controller.go:359]: I'm unsure what you mean, could you please explain more?

File: controllers/placementapi_controller.go
Comment (Deydra71) [reply controllers/placementapi_controller.go:415]: Thanks a lot for raising this and the example, I added it to the other tests.

File: api/go.mod
Comment (gibizer) [inline api/go.mod]: Hm, I'm wondering what pulling in newer k8s versions. Nothing should be.

File: controllers/placementapi_controller.go
Comment (gibizer) [reply controllers/placementapi_controller.go:415]: Thanks!

**Comment by gibizer** [comment]
/test placement-operator-build-deploy-kuttl
```
Warning: Push failed, retrying in 5s ...
Registry server Address: 
Registry server User Name: openstack-k8s-operators+cirobot
Registry server Email: 
Registry server Password: <<non-empty>>
error: build error: Failed to push image: trying to reuse blob sha256:07a64a71e01156f8f99039bc246149925c6d1480d3957de78510bbec6ec68f7a at destination: pinging container registry quay.rdoproject.org: Get "https://quay.rdoproject.org/v2/": net/http: TLS handshake timeout
```

**Review by gibizer — COMMENTED** [review]
I'm happy with this. So

File: api/go.mod
Comment (Deydra71) [reply api/go.mod]: These are unwanted remnants, let me remove them. I also forgot to remove the dependency to https://github.com/openstack-k8s-operators/openstack-operator/pull/625, so I edit it as well.

File: controllers/placementapi_controller.go
Comment (bogdando) [reply controllers/placementapi_controller.go:359]: Instead of calling ``mgr.GetFieldIndexer()`` multiple times for each [*Field getter](https://github.com/openstack-k8s-operators/placement-operator/pull/92/files#diff-c0f1c404ea10bc348fba06aab6f455b95b7513a337e09b325ebea615fc0ee587R329), you could call it to get the full spec then parse that local copy by the looked up getters.

If mgr.GetFieldIndexer() doesn't allow getting the full spec, you could get ``.spec.tls`` at least

File: controllers/placementapi_controller.go
Comment (bogdando) [reply controllers/placementapi_controller.go:1012]: meant doing the same as you did [above](https://github.com/openstack-k8s-operators/placement-operator/pull/92/files#diff-c0f1c404ea10bc348fba06aab6f455b95b7513a337e09b325ebea615fc0ee587R984)

**Review by bogdando — CHANGES_REQUESTED** [review]

File: controllers/placementapi_controller.go
Comment (Deydra71) [reply controllers/placementapi_controller.go:359]: I think I understand now your intention. The `IndexField()` function is not for retrieval, but for watching the changes, so storing the spec in a local var would compromise the reconciliation upon changes.

File: controllers/placementapi_controller.go
Comment (Deydra71) [reply controllers/placementapi_controller.go:1012]: I see, but the `deplDef` has not been declared before, apart from `ctrlResult` as used [above](https://github.com/openstack-k8s-operators/placement-operator/pull/92/files#diff-c0f1c404ea10bc348fba06aab6f455b95b7513a337e09b325ebea615fc0ee587R984).

File: controllers/placementapi_controller.go
Comment (bogdando) [reply controllers/placementapi_controller.go:359]: makes sense for the whole spec, thanks for explaining! what about storing just .spec.tls? That still saves a bunch of redundant requests to kube API server

Comment (Deydra71) [reply controllers/placementapi_controller.go:359]: That might work. I will need take a look if cached data is updated when the `.Spec.TLS` changes. Which will most probably require reconciler to compare the actual  `.Spec.TLS` against the stored one. 

CC @stuggi  what he might think about this.

Comment (stuggi) [reply controllers/placementapi_controller.go:359]: I don't think this will work, but lets investigate any optimization to index information in follow ups as discussed in https://github.com/openstack-k8s-operators/placement-operator/pull/92/files#r1457573645

**Review by gibizer — APPROVED** [review]
Looks good to me. I assume it was tested together with https://github.com/openstack-k8s-operators/openstack-operator/pull/625 and works.

File: controllers/placementapi_controller.go
Comment (bogdando) [reply controllers/placementapi_controller.go:359]: putting this solved to move discussions to the linked thread

Comment (bogdando) [reply controllers/placementapi_controller.go:359]: please make sure to create a follow up

Comment (stuggi) [reply controllers/placementapi_controller.go:359]: I have created https://issues.redhat.com/browse/OSPRH-4208 to keep tracking this

**Pull request merged**
